core/domain/caching_services.py

```diff
@@ -0,0 +1,125 @@
+# coding: utf-8
+#
+# Copyright 2014 The Oppia Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Service functions to set and retrieve data from the memory cache."""
+
+from __future__ import absolute_import  # pylint: disable=import-only-modules
+from __future__ import unicode_literals  # pylint: disable=import-only-modules
+
+from core.domain import collection_domain
+from core.domain import exp_domain
+from core.domain import skill_domain
+from core.domain import story_domain
+from core.domain import topic_domain
+from core.platform import models
+import python_utils
+
+memory_cache_services = models.Registry.import_cache_services()
+
+
+def flush_memory_cache():
+    """Flushes the redis cache by wiping all of the data."""
+    memory_cache_services.flush_cache()
+
+
+def _get_correct_type_of_key(key):
+    """In the memory cache, values are stored as (key, value) pairs where values
+    can be string representations of Oppia objects, e.g Collection,
+    Exploration, etc. These object types can be identified by the key that
+    the memory cache uses to store the dictionaries. This function returns the
+    correct type of the saved memory cache value using the key, so that the
+    encoded string can be decoded into an object.
+
+    Args:
+        key: str. The key string used in the memory cache.
+
+    Returns:
+        Collection|Exploration|Skill|Story|Topic|None. Returns the original
+        class of the object that got converted to a dictionary. If this key does
+        not correspond to a class that requires deserialization, return None.
+    """
+    if key.startswith('collection'):
+        return collection_domain.Collection
+    elif key.startswith('exp'):
+        return exp_domain.Exploration
+    elif key.startswith('skill'):
+        return skill_domain.Skill
+    elif key.startswith('story'):
+        return story_domain.Story
+    elif key.startswith('topic'):
+        return topic_domain.Topic
+    else:
+        return None
```

Comment:
So the problem with that is the current API sets the values as a key_value mapping which is a dictionary. This means that I can't relate the namespace to the individual keys unless we restrict each set_multi to be only setting one type instead of being able to set multiple types at once. Is this functionality that we should restrict? Each of the individual classes have their own key generation, i can change it to use a hyphen system. 

The reason I didn't raise an error is that anything that does not have that key will probably be a primitive type so we don't need to handle it all when retrieving or setting it to the cache. If this is not that clear, does adding a comment suffice?

Reply:
I think it's quite reasonable to have set_multi() only set one type at once. Is there any current usage in the codebase which conflicts with that? If not, then let's do the hyphen system -- it's cleaner.

For your second paragraph, see the second question in my first paragraph -- do we use memcache for storing type-less things? If so, let's have a prefix for that too. Otherwise if someone happens to store a typeless thing that has the same format as one of the typed things, confusion can result.